KAFKA-19862: Group coordinator loading may fail when there is concurrent compaction

izzyharker opened the pull request:
When the group coordinator loads with concurrent compaction, the record
unassigning a partition/task can be missed which leads to an
IllegalStateException being thrown, however the records self-resolve by
the time loading is finished. This PR makes the change to log a warning
and proceed with loading instead of throwing an IllegalStateException.

Testing: Added unit tests for Consumer and Streams groups to ensure this
change does not impact loading success.

Reviewers: Sean Quah <squah@confluent.io>, Lucas Brutschy
 <lbrutschy@confluent.io>, David Jacot <djacot@confluent.io>

izzyharker committed:
Removed IllegalStateException on partition reassignments

izzyharker committed:
removed unused import

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -77,6 +79,8 @@
  */
 public class ConsumerGroup extends ModernGroup<ConsumerGroupMember> {
 
+    private static final Logger log = LoggerFactory.getLogger(ConsumerGroup.class);
```
We have to use the `LogContext` from the `GroupMetadataManager`, otherwise we won't prefix log lines with `[GroupCoordinator id=%d]`. When adding `LogContext` to the constructor, it should go as the first parameter.

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1047,7 +1051,7 @@ void removePartitionEpochs(
                         Integer prevValue = partitionsOrNull.remove(partitionId);
                         if (prevValue != expectedEpoch) {
-                            throw new IllegalStateException(
+                            log.warn(
```
When we hit this case we must not have removed the entry from the map. We can replace the `remove` above with a `get` and `remove` if we don't hit this check.

The same for streams

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1034,7 +1038,7 @@ private void maybeRemovePartitionEpoch(
      * @param assignment    The assignment.
      * @param expectedEpoch The expected epoch.
-     * @throws IllegalStateException if the epoch does not match the expected one.
+     * @throws IllegalStateException if the epoch does not exist.
```
Thinking about this again, I think we can also hit the IllegalStateException case
```
Member A is assigned topic 1 partition X
[removed by compaction] Member A is unassigned topic 1 partition X
Member B is assigned topic 1 partition X
Member B is unassigned topic 1 partition X
Member A is assigned topic 2 partition Y
```

so we have to also turn that into log message and add another test.

The same for streams.

izzyharker replied:
Did you mean hitting the same IllegalStateException or the second check in the method? I made a second test with that order for the ConsumerGroup and didn't have any more issues.

squah-confluent replied:
I'm referring to the remaining IllegalStateException case `Cannot remove the epoch %d from %s because it does not have any epoch`

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24057,4 +24057,196 @@ private Map<Uuid, DeleteShareGroupOffsetsResponseData.DeleteShareGroupOffsetsRes
+            .build();
+
+        // Member A is assigned partition bar-0,1 and member B is assigned partition bar-2 
+        assignor.prepareGroupAssignment(new GroupAssignment(
```
We should be `replay`ing the records directly, as if we are loading the group coordinator. The records can be created using `GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord`.

The same for streams

izzyharker committed:
Updated logcontext in ConsumerGroup and partition update ordering

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -147,6 +149,29 @@ public String toLowerCaseString() {
 
     private final TimelineObject<Boolean> hasSubscriptionMetadataRecord;
 
+    private final LogContext logContext;
```
We can drop this field since we never use it.

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -147,6 +149,29 @@ public String toLowerCaseString() {
 
+    private final LogContext logContext;
+
+    private final Logger log;
```
nit: I'd move this to the top of the fields, before `TimelineObject<ConsumerGroupState> state`.

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -147,6 +149,29 @@ public String toLowerCaseString() {
+        this.resolvedRegularExpressions = new TimelineHashMap<>(snapshotRegistry, 0);
+        this.hasSubscriptionMetadataRecord = new TimelineObject<>(snapshotRegistry, false);
+        this.logContext = logContext;
+        this.log = logContext.logger(ConsumerGroup.class);
```
nit: same, I'd move this to the top of the constructor, after super().

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -147,6 +149,29 @@ public String toLowerCaseString() {
+
+    private final Logger log;
+
+    public ConsumerGroup(
+        LogContext logContext,
+        SnapshotRegistry snapshotRegistry,
+        String groupId
+    ) {
```
Can we remove the other constructor? In the tests we can fill in the extra parameter with `new LogContext()`.

izzyharker replied:
I think the second constructor can just call the first unless it really needs to be removed? That way it's still not as redundant.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24057,4 +24057,318 @@ private Map<Uuid, DeleteShareGroupOffsetsResponseData.DeleteShareGroupOffsetsRes
     }
+
+    @Test
+    public void testConsumerGroupResolvesOnCompaction() {
```
Is it possible to simpify the tests by removing the consumer group setup, target assignment records and heartbearts? To check what's in the `currentPartitionEpoch` map, we can assert on `consumerGroup.currentPartitionEpoch(topicId, partitionId)`.

izzyharker committed:
streams test added

izzyharker committed:
ConsumerGroup update

izzyharker committed:
Added second constructor

izzyharker committed:
fix

izzyharker committed:
Removed second IllegalStateException check for log msg

izzyharker committed:
Updated ConsumerGroup constructor

izzyharker committed:
removed non-logging constructor and fixed tests

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -103,6 +105,8 @@ public String toLowerCaseString() {
         }
     }
 
+    private final Logger log;
```
```suggestion
    /**
     * The logger.
     */
    private final Logger log;
```

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1121,6 +1128,7 @@ public ConsumerGroupDescribeResponseData.DescribedGroup asDescribedGroup(
     /**
      * Create a new consumer group according to the given classic group.
      *
+     * @param logContext        The LogContext.
```
```suggestion
     * @param logContext        The log context.
```

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -918,7 +918,7 @@ void removeTaskProcessIds(
      * @param assignment    The assignment.
      * @param expectedProcessId The expected process ID.
-     * @throws IllegalStateException if the process ID does not match the expected one. package-private for testing.
+     * @throws IllegalStateException if the process ID does not exist. package-private for testing.
```
```suggestion
     * package-private for testing.
```

izzyharker committed:
Simplified and updated tests.

izzyharker committed:
Updated syntax and formatting discrepancies from rebase.

izzyharker committed:
nit: tooltip fix

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1048,11 +1053,13 @@ void removePartitionEpochs(
+                        Integer prevValue = partitionsOrNull.get(partitionId);
                         if (prevValue != expectedEpoch) {
-                            throw new IllegalStateException(
+                            log.warn(
```
I am not a fan of those warnings. I think that users will be confused by them and they will think that something is wrong even if it is actually expected. Should we log them as debug? I would also include the group id. The message is meaningless without the group id.

izzyharker replied:
The LogContext gets passed from the GroupMetadataManager so the log messages should be prefixed with`[GroupCoordinator id=%d]` already.

izzyharker replied:
Logging the messages as debug is fine by me.

dajac replied:
Having `[GroupCoordinator id=%d]` is great but I still believe that having the group id is also needed in order to have a full context. In `GroupMetadataManager`, we usually add `[GroupId {}] ` after `[GroupCoordinator id=%d]`.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1089,7 +1096,7 @@ void addPartitionEpochs(
                 for (Integer partitionId : assignedPartitions) {
                     Integer prevValue = partitionsOrNull.put(partitionId, epoch);
```
While we are here, I wonder if we need to strengthen the logic. At minimum, we could ensure that a member can only "acquire the lock" if is has an epoch larger than the previous one. I am not sure whether we could have records reordered to be in this situation though. @squah-confluent What do you think?

izzyharker replied:
Now that you mention it, I think it could be good to at least update the logic to be more in line with the `remove` methods -so `get` the value first then only `putting` it if we don't hit the check.

squah-confluent replied:
I think compaction only allows records to be dropped and not reordered (they need to keep their offsets!). So we should only replace a non-null epoch with a larger one here and it makes sense to add a check to throw an IllegalStateException otherwise.

dajac reviewed (commented):
@izzyharker Thanks for the fix. I left a few comments for consideration.

izzyharker committed:
Swapped log.warn for log.debug

izzyharker committed:
Fixed ConsumerGroup/StreamsGroup tests, added groupId to log

squah-confluent reviewed (commented):
I've only reviewed the non-test parts for now. I'll get to the tests soon-ish.

izzyharker committed:
Comment fix and logic update for addPartitionEpochs

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1048,11 +1058,13 @@ void removePartitionEpochs(
                     assignedPartitions.forEach(partitionId -> {
-                        Integer prevValue = partitionsOrNull.remove(partitionId);
+                        Integer prevValue = partitionsOrNull.get(partitionId);
                         if (prevValue != expectedEpoch) {
```
nit: Let's inverse the condition so we have the happy path first.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1087,12 +1100,16 @@ void addPartitionEpochs(
-                        throw new IllegalStateException(
-                            String.format("Cannot set the epoch of %s-%s to %d because the partition is " +
-                                "still owned at epoch %d", topicId, partitionId, epoch, prevValue));
+                        if (prevValue > epoch) {
```
nit: Let's put the happy path first here too and we can also combine the two if conditions into one.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -942,11 +942,13 @@ private void removeTaskProcessIds(
                     assignedPartitions.keySet().forEach(partitionId -> {
-                        String prevValue = partitionsOrNull.remove(partitionId);
+                        String prevValue = partitionsOrNull.get(partitionId);
                         if (!Objects.equals(prevValue, expectedProcessId)) {
```
Same comments for this file.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,263 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(topicId, 0)))
+            .build()));
+
+        // Assign partition 0 to member B. This is allowed even though partition 0 is already owned by member A.
```
Let's explain why this is possible. We could perhaps add a comment with all the records that would be written and how compaction could remove some of those? My concern is that it is not an obvious issue to grasp for non-experts.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,263 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+
+        // Verify partition epochs.
+        assertEquals(group.currentPartitionEpoch(topicId, 0), 12);
+        assertEquals(group.currentPartitionEpoch(topicId, 1), 13);
```
nit: The expected value must be first and the actual value second. This applies to all the assertions added in this PR.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,263 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+    }
+
+    @Test
+    public void testConsumerGroupUnassignmentResolvesWithCompaction() {
```
I tried this test _without the fix_ and it passes!?

squah-confluent replied:
I don't think the test setup is quite right. The original intention was to exercise the branch of removePartitionEpochs where the topic has no epochs at all. Seeing as we already test that in ConsumerGroupTest (or will once the other review comments are resolved), perhaps we can remove this test case.

izzyharker replied:
The initial variant of this test had all the records replayed (the unassignment included) so it didn't actually check the compaction scenario with a removed unassignment record. Removing that replay gave the expected test behavior. I agree that we could perhaps remove the test though.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,263 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+
+        ConsumerGroup group = context.groupMetadataManager.consumerGroup(groupId);
+
+        // Assign partition foo-1 to member A
```
nit: Add a `.` at the end of the sentence to be consistent.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,263 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setPreviousMemberEpoch(10)
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(fooTopicId, 0, 1)))
+            .build()));
+
```
nit: We can remove an empty line.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -149,10 +157,13 @@ public String toLowerCaseString() {
     ) {
         super(snapshotRegistry, groupId);
+        // Add the GroupId to the log prefix for context.
+        this.log = new LogContext(String.format("%s [Group %s]: ", logContext.logPrefix(), groupId)).logger(ConsumerGroup.class);
```
I just recalled about https://github.com/apache/kafka/pull/19138 so I would prefer to inline adding the prefix given that we have only a few logs in this class. It is probably OK like this but I prefer to be on the safe side.

dajac commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -942,11 +942,13 @@ private void removeTaskProcessIds(
                         if (!Objects.equals(prevValue, expectedProcessId)) {
-                            throw new IllegalStateException(
+                            log.debug(
                                 String.format("Cannot remove the process ID %s from task %s_%s because the partition is " +
```
We need to prefix those too.

dajac reviewed (commented):
@izzyharker Thanks for the update. I left some comments for consideration.

izzyharker committed:
Changed the following:
- Updated test conditions to accurately demo compaction playback issues
- Swapped logic order to have the happy path first in StreamsGroup & ConsumerGroup
- Fixed inconsistencies in comments
- Updated log messages to include GroupId inline instead of re-initializing log for performance

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1048,11 +1057,14 @@ void removePartitionEpochs(
+                            // GroupId added for context. 
+                            log.debug(
+                                String.format("[Group %s]: Cannot remove the epoch %d from %s-%s because the partition is " +
+                                    "still owned at a different epoch %d", groupId, expectedEpoch, topicId, partitionId, prevValue));
```
* We can use the logger's string formatting capabilities.
* Elsewhere, we prefix with "[GroupId {}] " ('Id' and no ':').

The same for the other log messages.

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1048,11 +1057,14 @@ void removePartitionEpochs(
+                        if (prevValue != null && prevValue == expectedEpoch) {
+                            partitionsOrNull.remove(partitionId);
+                        } else {
+                            // GroupId added for context. 
```
```suggestion
```
Can we drop this comment? The same for streams.

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroup.java:
```diff
@@ -1087,8 +1100,10 @@ void addPartitionEpochs(
-                    Integer prevValue = partitionsOrNull.put(partitionId, epoch);
-                    if (prevValue != null) {
+                    Integer prevValue = partitionsOrNull.get(partitionId);
+                    if (prevValue == null || prevValue <= epoch) {
```
Can we bring back the rejection for replacing the epoch with an identical one?
I don't think it should ever happen unless I'm missing something.
```suggestion
                    if (prevValue == null || prevValue < epoch) {
```

squah-confluent replied:
Actually I was wrong. It's possible to see a replacement with the same epoch:

1. Member A unsubscribes from `bar` at epoch N - 1. `bar` is pending revocation due to the fix for KAFKA-19431. `Member A { epoch: N - 1, assigned partitions: [foo], pending revocation: [bar] }`
2. A new assignment is available with epoch N.
3. Member A yields bar. The epoch is bumped. `Member A { epoch: N, assigned partitions: [], pending revocation: [foo] }`
4. Member A yields foo. `Member A { epoch: N, assigned partitions: [], pending revocation: []}`
5. Member B is assigned foo. `Member B { epoch: N, assigned partitions: [foo] }`

When the record from 4 is dropped by compaction, foo's partition epoch is replaced with an identical epoch.

We should also document this in a test. It can be called something like `testReplayConsumerGroupCurrentMemberAssignmentSameEpochWithCompaction`

dajac replied:
Step 4. does not seem correct to me in this example as I explained in the comment about the test itself.

dajac replied:
We can revert this change and remove the test as this is addressed by https://github.com/apache/kafka/pull/21107.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroupTest.java:
```diff
@@ -288,22 +289,17 @@ public void testUpdatingMemberUpdatesPartitionEpochWhenPartitionIsNotReleased()
         Uuid fooTopicId = Uuid.randomUuid();
         ConsumerGroup consumerGroup = createConsumerGroup("foo");
 
-        // Removing should fail because there is no epoch set.
-        assertThrows(IllegalStateException.class, () -> consumerGroup.removePartitionEpochs(
-            mkAssignment(
-                mkTopicAssignment(fooTopicId, 1)
-            ),
-            10
-        ));
-
```
Can we keep this? It just won't throw.
```
// Removing should be a no-op when there is no epoch set.
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroupTest.java:
```diff
@@ -312,13 +308,19 @@ public void testRemovePartitionEpochs() {
 
-        // Removing should fail because the expected epoch is incorrect.
-        assertThrows(IllegalStateException.class, () -> consumerGroup.removePartitionEpochs(
+        // Removing with incorrect epoch should fail. 
```
```suggestion
        // Removing with incorrect epoch should do nothing. 
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroupTest.java:
```diff
@@ -288,22 +289,17 @@ public void testUpdatingMemberUpdatesPartitionEpochWhenPartitionIsNotReleased()
+        consumerGroup.updateMember(m2);
+        assertEquals(mkAssignment(mkTopicAssignment(fooTopicId, 1)), 
+            consumerGroup.getOrMaybeCreateMember("m1", false).assignedPartitions()
+        );
```
This member update ought to be rejected (throw an IllegalStateException like previously expected).

squah-confluent replied:
^ Ignore the above.

We should however test that an `IllegalStateException` is thrown when the epoch goes backwards and the update is accepted when the epoch is the same.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -375,24 +375,20 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
         String fooSubtopologyId = "foo-sub";
         StreamsGroup streamsGroup = createStreamsGroup("foo");
 
-        // Removing should fail because there is no epoch set.
-        assertThrows(IllegalStateException.class, () -> streamsGroup.removeTaskProcessIds(
-            TaskAssignmentTestUtil.mkTasksTupleWithCommonEpoch(taskRole, 10, mkTasks(fooSubtopologyId, 1)),
-            "process"
-        ));
-
```
Can we keep this?
```
// Removing should be a no-op when there is no epoch set.
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -375,24 +375,20 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
 
-        // Removing should fail because the expected epoch is incorrect.
-        assertThrows(IllegalStateException.class, () -> streamsGroup.removeTaskProcessIds(
+        // Removing with incorrect epoch should fail. 
```
```suggestion
        // Removing with incorrect epoch should do nothing. 
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -375,24 +375,20 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
             "process1"
-        ));
+        );
+        assertEquals(m1.assignedTasks(), streamsGroup.getMemberOrThrow("m1").assignedTasks());
```
We're missing an assert for the epoch itself.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,268 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .withConfig(GroupCoordinatorConfig.CONSUMER_GROUP_ASSIGNORS_CONFIG, List.of(assignor))
+            .withConsumerGroup(new ConsumerGroupBuilder(groupId, 10)
+                .withMetadataHash(topicHash))
+            .build();
```
We can remove most of this setup and let the `replay` method create the group. The group lookup will have to be moved to after the `replay`s.

The same for streams.
```suggestion
        CoordinatorMetadataImage metadataImage = new MetadataImageBuilder()
            .addTopic(topicId, topicName, 2)
            .buildCoordinatorMetadataImage();

        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
            .withMetadataImage(metadataImage)
            .build();
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,268 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setMemberEpoch(13)
+            .setPreviousMemberEpoch(12)
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(topicId, 1)))
+            .build()));
```
```suggestion
            .build()));
        // Partition 0 must remain with member B at epoch 12 even though member A has just been unassigned partition 0.
```

+equivalent comment for the streams version.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,268 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(topicId, 0)))
+            .build()));
+
+        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberB)
```
```suggestion
        // Partition 0's owner is replaced with member B at epoch 12.
        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberB)
```

+equivalent comment for the streams version.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,268 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
     }
 
+    @Test
+    public void testConsumerGroupAssignmentResolvesWithCompaction() {
```
How about
```suggestion
    public void testConsumerGroupPartitionEpochsWithCompaction() {
```
The same for streams.

squah-confluent replied:
`testReplayConsumerGroupCurrentMemberAssignmentWithCompaction` is probably a better name.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,268 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setState(MemberState.STABLE)
+            .setMemberEpoch(14)
+            .setPreviousMemberEpoch(13)
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(fooTopicId, 2)))
```
We need to clear out all `foo` epochs here.

squah-confluent replied:
We missed `foo-0`. It's still owned by the other member.

squah-confluent reviewed (commented):
Thanks for updating the PR! I finally got around to reviewing the tests.

izzyharker committed:
- Reformatted log messages
- Added test for pending revocation assignments
- Refined tests

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .withMetadataImage(new MetadataImageBuilder()
+            .addTopic(topicId, topicName, 2)
+            .buildCoordinatorMetadataImage())
```
The indentation's not quite right here.
```suggestion
                .addTopic(topicId, topicName, 2)
                .buildCoordinatorMetadataImage())
```

The same for the other tests.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(topicId, 0)))
+            .build()));
+
+        // Partition 0 must remain with member B at epoch 12 even though member A just been unassigned partition 0.
```
typo: I think we're missing a word here
```suggestion
        // Partition 0 must remain with member B at epoch 12 even though member A has just been unassigned partition 0.
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        // 5. Member A is assigned partition bar-0. 
+        // This is a legitimate set of assignments but with compaction the unassignment record can be skipped.
+        // We would like to not fail in these cases and allow both the assignment of member B to foo-1 and 
+        // member A to bar-0 to succeed because the epochs are larger. 
```
The idea is that at step 5, there are supposed to be no epochs for any `foo` partition so that we hit the `[GroupId {}] Cannot remove the epoch {} from {} because it does not have any epoch` path. Let's either fix this test (and make it clear in this comment).

The same for the streams version.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        StreamsGroup group = context.groupMetadataManager.streamsGroup(groupId);
+        assertEquals(TaskAssignmentTestUtil.mkTasksTupleWithEpochs(TaskRole.ACTIVE, 
+            TaskAssignmentTestUtil.mkTasksWithEpochs(subtopologyBar, Map.of(0, 13))), group.members().get(memberA).assignedTasks());
+    }
```
```suggestion
    }

```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        assertEquals(TaskAssignmentTestUtil.mkTasksTupleWithEpochs(TaskRole.ACTIVE, 
+            TaskAssignmentTestUtil.mkTasksWithEpochs(subtopology, Map.of(1, 13))), group.members().get(memberA).assignedTasks());
+        assertEquals(TaskAssignmentTestUtil.mkTasksTupleWithEpochs(TaskRole.ACTIVE, 
+            TaskAssignmentTestUtil.mkTasksWithEpochs(subtopology, Map.of(0, 12))), group.members().get(memberB).assignedTasks());
```
Can we assert on `group.currentActiveTaskProcessId` instead? `member.assignedTasks` will be trivially correct.
```suggestion
        assertEquals(memberB, group.currentActiveTaskProcessId(subtopology, 0));
        assertEquals(memberA, group.currentActiveTaskProcessId(subtopology, 1));
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+    public void testReplayConsumerGroupCurrentMemberAssignmentWithCompaction() {
+        String groupId = "fooup";
+        String memberA = "memberA";
+        String memberB = "memberB";
```
Could we call these `memberIdA` and `memberIdB`?
```suggestion
        String memberIdA = "memberA";
        String memberIdB = "memberB";
```

The same for the other tests.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,296 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+                    TaskAssignmentTestUtil.mkTasksWithEpochs(subtopology, Map.of(0, 11))))
+            .build()));
+
+        // Task 0's owner is replaced by member B at epoch 12.
+        context.replay(StreamsCoordinatorRecordHelpers.newStreamsGroupCurrentAssignmentRecord(groupId, streamsGroupMemberBuilderWithDefaults(memberB)
+            .setMemberEpoch(12)
+            .setPreviousMemberEpoch(11)
+            .setAssignedTasks(TaskAssignmentTestUtil.mkTasksTupleWithEpochs(TaskRole.ACTIVE, 
+                    TaskAssignmentTestUtil.mkTasksWithEpochs(subtopology, Map.of(0, 12))))
+            .build()));
```
I don't understand why we aren't throwing an exception here, because we're missing the fix for `addTaskProcessIdFromActiveTasksWithEpochs`. Can you spot what I'm missing here?

izzyharker replied:
~~Still working on figuring out the cause but it seems like the`currentActiveTaskToProcessId` map isn't getting populated correctly for some reason. When I swapped the assert to use `group.currentActiveTaskProcessId` on the Streams tests it keeps giving `null` even though the `group.members` fields are correct.~~

Turns out you have to initialize the streams group members with metadata otherwise some of the fields don't get populated and the updates don't actually go through. Much more confusing than the consumer group interface for sure.

lucasbru replied:
Can you explain this in more detail? Initializing the field of a streams group member `null` indeed is a mistake that would be nice to get fixed. Just want to check if there is a bug lurking here.

izzyharker replied:
We're using `StreamsCoordinatorRecordHelpers.newStreamsGroupCurrentAssignmentRecord` for the test cases in order to replay records checking different scenarios where compaction may drop some unassignments. If you create and only interact with a group member using that assignment function, the process-id field of the member just gets `null` and so the members get added to the `members` field in the group but the `ActiveTasks` map doesn't get updated (which was the testing issue). 

I don't think it's a bug so much as a misunderstanding of how StreamsGroup members are set up from my side because I was basing it off of the tests for ConsumerGroup members where it's not necessary to take that extra step.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroupTest.java:
```diff
@@ -286,23 +287,37 @@ public void testUpdatingMemberUpdatesPartitionEpochWhenPartitionIsNotReleased()
+        consumerGroup.updateMember(m2);
+        assertEquals(mkAssignment(mkTopicAssignment(fooTopicId, 1)), 
+            consumerGroup.getOrMaybeCreateMember("m1", false).assignedPartitions()
+        );
```
I can't follow the purpose of this assert? (It looks like `getOrMaybeCreateMember` will always return the exact same `ConsumerGroupMember` instance passed into `updateMember`).

izzyharker replied:
The `getOrMaybeCreateMember` is used for assertions in a number of the other tests as well, it was meant to be a check that the `removePartitionEpochs` call didn't change the member assignment. It's maybe not necessary to have both that and the assert on the epoch though.

squah-confluent replied:
I'm strongly in favor of removing the asserts on `member.assignedPartitions()` in the epoch/process id-based tests. The partition epoch methods aren't expected to ever touch the member assignment, since the Member objects are designed to be immutable.

squah-confluent reviewed (commented):
Thanks for updating the PR! We're getting closer.

I realized that we didn't update the streams `addTaskProcessIdFromActiveTasksWithEpochs`, so I'm confused about why the streams tests are passing.

izzyharker committed:
- Applied fix to StreamsGroup#addTaskProcessIdFromActiveTasksWithEpochs
- Fixed test bug in streams tests where updates weren't being applied
- Small formatting nits

izzyharker commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -1027,12 +1027,28 @@ private void addTaskProcessIdFromActiveTasksWithEpochs(
+                            throw new IllegalStateException(
+                                String.format("[GroupId {}] Cannot set the process ID of {}-{} to {} because the partition is " +
+                                "still owned by process ID {}", groupId, subtopologyId, partitionId, processId, prevValue));
+                        }
```
Applied the same version of the fix as in ConsumerGroup#addPartitionEpochs. The way epochs are implemented for the Streams groups is not very conducive to this use case which is why it's a bit convoluted.

squah-confluent replied:
This seems like a huge hassle to iterate through all the members of the group. When groups have lots of members it could become a bottleneck. I would propose dropping the check for epochs going backwards for streams.

lucasbru replied:
Yes, we are using processIds here to guide the assignment (no stateful tasks can be assigned to the same processID), but we don't identify a single member. That could be added, of course, but we saw that this wasn't really used in consumer groups. So I would support not doing this check here.

izzyharker replied:
Thanks!

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -375,11 +374,11 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
 
-        // Removing should fail because there is no epoch set.
-        assertThrows(IllegalStateException.class, () -> streamsGroup.removeTaskProcessIds(
+        // Removing should be a no-op when there is no epoch set.
```
```suggestion
        // Removing should be a no-op when there is no process id set.
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -388,11 +387,17 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
-        // Removing should fail because the expected epoch is incorrect.
-        assertThrows(IllegalStateException.class, () -> streamsGroup.removeTaskProcessIds(
-            TaskAssignmentTestUtil.mkTasksTupleWithCommonEpoch(taskRole, 10, mkTasks(fooSubtopologyId, 1)),
+        // Removing with incorrect epoch should do nothing. 
```
```suggestion
        // Removing with incorrect process id should do nothing. 
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -388,11 +387,17 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
+        // Removing with incorrect epoch should do nothing. 
+        // A debug message is logged, no exception is thrown.
+        streamsGroup.removeTaskProcessIds(
+            TaskAssignmentTestUtil.mkTasksTupleWithCommonEpoch(taskRole, 9, mkTasks(fooSubtopologyId, 1)),
             "process1"
-        ));
+        );
```
Can we assert that the process id is unchanged?
```java
if (taskRole == TaskRole.ACTIVE) {
    assertEquals("process", streamsGroup.currentActiveTaskProcessId(fooSubtopologyId, 1));
}
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+                    TaskAssignmentTestUtil.mkTasksWithEpochs(subtopologyId, Map.of(0, 12))))
+            .build()));
+
+        // Task 0 must remain with member B at epoch 12 even though member A has just been unassigned task 1.
```
```suggestion
        // Task 0 must remain with member B at epoch 12 even though member A has just been unassigned task 0.
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        // 5. Member A is assigned partition bar-0. 
+        // This is a legitimate set of assignments but with compaction the unassignment record can be skipped.
+        // This can lead to conflicts from updating an owned partition in step 3 and attempting to remove
+        // nonexistent ownership in step 5. We want to ensure both assignments are allowed.  
```
We're not really interested in step 3 because that's covered by the test case above.
```suggestion
        // This can lead to conflicts from updating an owned partition in step 3 and attempting to remove
        // nonexistent ownership in step 5. We want to ensure that removing ownership from a
        // completely unowned topic in step 5 is allowed.
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+    }
+
+    @Test
+    public void testReplayConsumerGroupUnassignmentRecordSkippedWithCompaction() {
```
```suggestion
    public void testReplayConsumerGroupCurrentMemberAssignmentUnownedTopicWithCompaction() {
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        // 1. Member A is assigned partition foo-1.
+        // 2. Member A is unassigned partition foo-1 [record removed by compaction].
+        // 3. Member B is assigned partition foo-1.
+        // 4. Member B is unassigned partition foo-1. 
```
We actually use `foo-0` below.
```suggestion
        // 1. Member A is assigned partition foo-0.
        // 2. Member A is unassigned partition foo-0 [record removed by compaction].
        // 3. Member B is assigned partition foo-0.
        // 4. Member B is unassigned partition foo-0. 
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(fooTopicId, 0)))
+            .build()));
+
+        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberIdB)
```
```suggestion
        // foo-0's owner is replaced by member B at epoch 12.
        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberIdB)
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setAssignedPartitions(mkAssignment(mkTopicAssignment(fooTopicId, 0)))
+            .build()));
+
+        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberIdB)
```
```suggestion
        // foo becomes unowned.
        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberIdB)
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setPreviousMemberEpoch(12)
+            .build()));
+
+        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberIdA)
```
```suggestion
        // Member A is unassigned foo-0.
        context.replay(GroupCoordinatorRecordHelpers.newConsumerGroupCurrentAssignmentRecord(groupId, new ConsumerGroupMember.Builder(memberIdA)
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        // Verify member A only has ownership of partition bar-0. Member B has no partitions.
+        ConsumerGroup group = context.groupMetadataManager.consumerGroup(groupId);
+        assertEquals(mkAssignment(mkTopicAssignment(barTopicId, 0)), group.members().get(memberIdA).assignedPartitions());
+        assertEquals(14, group.currentPartitionEpoch(barTopicId, 0));
```
For completeness, let's check foo-0 as well.
```suggestion
        assertEquals(-1, group.currentPartitionEpoch(fooTopicId, 0));
        assertEquals(14, group.currentPartitionEpoch(barTopicId, 0));
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+    }
+
+    @Test
+    public void testReplayStreamsGroupUnassignmentRecordSkippedWithCompaction() {
```
```suggestion
    public void testReplayStreamsGroupCurrentMemberAssignmentUnownedTopologyWithCompaction() {
```

And could you make the structure of this test mirror the consumer groups one? They're out of sync.

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+            .setMemberEpoch(11)
+            .setPreviousMemberEpoch(10)
+            .setAssignedTasks(TaskAssignmentTestUtil.mkTasksTupleWithEpochs(TaskRole.ACTIVE, 
+                TaskAssignmentTestUtil.mkTasksWithEpochs(subtopologyFoo, Map.of(0, 11, 1, 11))))
```
```suggestion
                TaskAssignmentTestUtil.mkTasksWithEpochs(subtopologyFoo, Map.of(0, 11))))
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,300 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+
+        // Check task bar-0 is assigned to member A. Member B has no tasks.
+        StreamsGroup group = context.groupMetadataManager.streamsGroup(groupId);
+        assertEquals(processIdA, group.currentActiveTaskProcessId(subtopologyBar, 0));
```
Let's check both tasks.
```suggestion
        assertEquals(null, group.currentActiveTaskProcessId(subtopologyFoo, 0));
        assertEquals(processIdA, group.currentActiveTaskProcessId(subtopologyBar, 0));
```

squah-confluent reviewed (commented):
Nice work getting to the bottom of the streams process id issue! The fix you added to the streams tests looks good.

Some tidy ups:

izzyharker committed:
- Removed epoch check from StreamsGroup#addTaskProcessIdFromActiveTasksWithEpochs
- Updated comments and formatting
- Made tests more accurate

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -1004,7 +1004,7 @@ private void removeTaskProcessIdsFromSet(
      * @param tasks     The assigned tasks.
      * @param processId The process ID.
-     * @throws IllegalStateException if the partition already has an epoch assigned. package-private for testing.
+     * @throws IllegalStateException if the existing partition has larger epoch than the new one. package-private for testing.
```
I think we don't compare epochs for streams now?
```suggestion
     * package-private for testing.
```

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -1027,12 +1027,11 @@ private void addTaskProcessIdFromActiveTasksWithEpochs(
                     partitionsOrNull = new TimelineHashMap<>(snapshotRegistry, assignedTaskPartitionsWithEpochs.size());
                 }
-                for (Integer partitionId : assignedTaskPartitionsWithEpochs.keySet()) {
+                for (Integer partitionId: assignedTaskPartitionsWithEpochs.keySet()) {
```
nit: missing space
```suggestion
                for (Integer partitionId : assignedTaskPartitionsWithEpochs.keySet()) {
```

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -1027,12 +1027,11 @@ private void addTaskProcessIdFromActiveTasksWithEpochs(
-                            String.format("Cannot set the process ID of %s-%s to %s because the partition is " +
-                                "still owned by process ID %s", subtopologyId, partitionId, processId, prevValue));
+                        log.debug("[GroupId {}] Cannot set the process ID of {}-{} to {} because the partition is " +
+                            "still owned by process ID {}", groupId, subtopologyId, partitionId, processId, prevValue);
```
Technically we are actually setting the process ID. Let's update the log message to be more accurate.
```suggestion
                        log.debug("[GroupId {}] Setting the process ID of {}-{} to {} even though the partition is " +
                            "still owned by process ID {}", groupId, subtopologyId, partitionId, processId, prevValue);
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroupTest.java:
```diff
@@ -312,13 +325,19 @@ public void testRemovePartitionEpochs() {
+        assertEquals(
+            mkAssignment(mkTopicAssignment(fooTopicId, 1)),
+            consumerGroup.getOrMaybeCreateMember("m1", false).assignedPartitions()
+        );
```
```suggestion
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -388,11 +389,16 @@ public void testRemoveTaskProcessIds(TaskRole taskRole) {
             "process1"
-        ));
+        );
+        assertEquals(m1.assignedTasks(), streamsGroup.getMemberOrThrow("m1").assignedTasks());
```
```suggestion
```

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/streams/StreamsGroupTest.java:
```diff
@@ -409,16 +415,17 @@ public void testAddTaskProcessIds() {
-        ));
+        );
+
+        assertEquals("process", streamsGroup.currentActiveTaskProcessId(fooSubtopologyId, 1));
```
Let's test the process ID replacement branch.
```suggestion
            "process1"
        );

        // We allow replacing with a different process id. 
        streamsGroup.addTaskProcessId(
            new TasksTupleWithEpochs(
                mkTasksPerSubtopologyWithCommonEpoch(10, mkTasks(fooSubtopologyId, 1)),
                mkTasksPerSubtopology(mkTasks(fooSubtopologyId, 2)),
                mkTasksPerSubtopology(mkTasks(fooSubtopologyId, 3))
            ),
            "process2"
        );

        assertEquals("process2", streamsGroup.currentActiveTaskProcessId(fooSubtopologyId, 1));
```

squah-confluent reviewed (commented):
Thanks for updating the PR. We're getting close now.

izzyharker committed:
- Fixed comments/tooltips
- Test updates

lucasbru commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -980,9 +981,8 @@ private void removeTaskProcessIdsFromSet(
                 if (partitionsOrNull != null) {
                     assignedPartitions.forEach(partitionId -> {
                         if (!partitionsOrNull.get(partitionId).remove(processIdToRemove)) {
```
```suggestion
                        if (!partitionsOrNull.containsKey(parititionId) || !partitionsOrNull.get(partitionId).remove(processIdToRemove)) {
```

lucasbru reviewed (commented):
Mostly looking good to me, left one comment.

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,287 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+        // 2. A new assignment is available at epoch 11 with member A unsubscribing from topic foo.
+        // 3. Member A yields bar. The epoch is bumped to 11: Member A { epoch: 11, assigned partitions: [], pending revocations: [foo] }
+        // 4. Member A yields topic foo. Member A { epoch: 11, assigned partitions: [], pending revocations: [] } [removed by compaction]
+        // 5. Member B is assigned topic foo. Member B { epoch: 11, assigned partitions: [foo], pending revocations: [] }
```
I am confused by this example. I thought that one of the core invariant of the protocol is that a partition cannot be owned by two members at the same epoch. When member A unsubscribe from foo, the group epoch must be bumped to trigger a new computation of the target assignment. foo will only be reassignment to a new member is this new epoch. Hence here member B should be at epoch 12. Am I missing something?

squah-confluent replied:
> I thought that one of the core invariant of the protocol is that a partition cannot be owned by two members at the same epoch.

I broke this in KAFKA-19431 Stronger assignment consistency with subscription for consumer groups without realizing at the time. The problem is that we bump the member epoch by 1 once its current set of partitions pending revocation are revoked. With KAFKA-19431, we have no way to distinguish between partitions pending revocation due to a new assignment vs due to an unsubscription and we bump the member epoch after revocation in both cases.

> When member A unsubscribe from foo

Member A doesn't do this. Epoch 11's assignment is triggered by the unsubscribe from bar and unassigns foo from member A.

dajac replied:
This is addressed as part of https://github.com/apache/kafka/pull/21107.

izzyharker replied:
With that change, are we looking to revise the criteria in `ConsumerGroup#addPartitionEpochs` to require a strictly larger epoch when updating (currently requires at least as large)?

squah-confluent replied:
Yes

izzyharker replied:
Do we still need this test then? It seems like the other two tests cover the cases unless I'm missing some understanding about the different scenarios.

squah-confluent replied:
The test is no longer required. It can be removed

dajac commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java:
```diff
@@ -24387,6 +24387,287 @@ public void testShareGroupRecordsNoExceptionOnReplay(CoordinatorRecord record) {
+
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder().build();
+
+        // This test enacts the following scenario:
```
I have two high level remarks regarding the example:
1) We are not consistent in the format between them. This is kind of annoying while reviewing.
2) I wonder if we should make them more precise and really follow the correct sequence of the protocol and mention all the steps (e.g. group epoch bumped, new target assignment, etc.) based on heartbeats. This particular example seems incomplete.

izzyharker replied:
When you say not consistent in the format, do you mean the wording of the comment/test explanation or is it something else? Just want to understand exactly what the issue is from your side for the future.

dajac replied:
> do you mean the wording of the comment/test explanation
Yes

squah-confluent commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/streams/StreamsGroup.java:
```diff
@@ -1030,9 +1030,8 @@ private void addTaskProcessIdFromActiveTasksWithEpochs(
-                        throw new IllegalStateException(
-                            String.format("Cannot set the process ID of %s-%s to %s because the partition is " +
-                                "still owned by process ID %s", subtopologyId, partitionId, processId, prevValue));
+                        log.debug("[GroupId {}]Setting the process ID of {}-{} to {} even though the partition is " +
```
```suggestion
                        log.debug("[GroupId {}] Setting the process ID of {}-{} to {} even though the partition is " +
```

izzyharker committed:
- Reverted same epoch change and removed test
- log update

squah-confluent commented on group-coordinator/src/test/java/org/apache/kafka/coordinator/group/modern/consumer/ConsumerGroupTest.java:
```diff
@@ -333,14 +348,24 @@ public void testAddPartitionEpochs() {
+        );
+        assertEquals(11, consumerGroup.currentPartitionEpoch(fooTopicId, 1));
+
+        // Updating to a smaller epoch should fail.
```
Now that we are rejecting the equal epoch case again, could we add a part of the test that tries to update to the same epoch and fails?

dajac reviewed (approved):
lgtm

dajac commented:
Merged to trunk and 4.2. I could not cherry-pick it to 4.1 and 4.0. @izzyharker Could you please raise a PR for those branches? I think that it is worth fixing this for future minor releases.
